packages/core/src/indexing/postgres/__tests__/postgres-index-api.test.ts

```diff
@@ -70,66 +81,177 @@ afterAll(async () => {
-          nullable: false,
-          defaultValue: null,
-        },
-      }
+      await expect(indexApi.verifyTables(indexModelsArgs)).resolves.not.toThrow()
+
+      // Also manually check table structure
+      const columns = await dbConnection.table(asTableName(indexModelsArgs[0].model)).columnInfo()
+      const expectedTableStructure = Object.assign({}, COMMON_TABLE_STRUCTURE, {
+        fooRelation: RELATION_COLUMN_STRUCTURE,
+      })
+      expect(JSON.stringify(columns)).toEqual(JSON.stringify(expectedTableStructure))
+    })
 
-      const modelsToIndex = [StreamID.fromString(STREAM_ID_A)]
+    test('table creation is idempotent', async () => {
+      const modelsToIndex = [Model.MODEL, StreamID.fromString(STREAM_ID_A)]
       const indexApi = new PostgresIndexApi(dbConnection, true, logger)
-      await indexApi.indexModels(modelsToIndex)
+      await indexApi.indexModels(modelsToIndexArgs(modelsToIndex))
+      // Index the same models again to make sure we don't error trying to re-create the tables
+      await indexApi.indexModels(modelsToIndexArgs(modelsToIndex))
       const created = await listMidTables(dbConnection)
       const tableNames = modelsToIndex.map(asTableName)
       expect(created).toEqual(tableNames)
-      await expect(indexApi.verifyTables(modelsToIndex, INVALID_TABLE_STRUCTURE)).rejects.toThrow(
-        /Schema verification failed for index/
-      )
     })
 
     test('create new table with existing ones', async () => {
       // First init with one model
       const modelsA = [StreamID.fromString(STREAM_ID_A)]
       const indexApiA = new PostgresIndexApi(dbConnection, true, logger)
-      await indexApiA.indexModels(modelsA)
+      await indexApiA.indexModels(modelsToIndexArgs(modelsA))
       const createdA = await listMidTables(dbConnection)
       const tableNamesA = modelsA.map(asTableName)
       expect(createdA).toEqual(tableNamesA)
 
       // Next add another one
       const modelsB = [...modelsA, StreamID.fromString(STREAM_ID_B)]
       const indexApiB = new PostgresIndexApi(dbConnection, true, logger)
-      await indexApiB.indexModels(modelsB)
+      await indexApiB.indexModels(modelsToIndexArgs(modelsB))
       const createdB = await listMidTables(dbConnection)
       const tableNamesB = modelsB.map(asTableName)
       createdB.sort()
       tableNamesB.sort()
       expect(createdB).toEqual(tableNamesB)
     })
   })
+
+  describe('table validation tests', () => {
+    /**
+     * This test exists mostly to validate the rest of the validation tests.
+     * Those tests test that validation fails in certain cases by changing one thing
+     * off from what we expect to pass validation.  This test confirms that we are in fact
+     * creating tables in this unit test in the correct way that could pass validation if not
+     * for the intentional changes we make in the following test.
+     *
+     * NOTE: If this test ever fails and needs updating, update the other validation tests
```

Comment:
This is pretty brittle and I don't love it.  I had some time while working on these tests where the test was passing because validation was failing, but it was failing for the wrong reason and so wasn't really testing what I thought it was.  I happened to notice and figure it out, but it seems pretty brittle and likely that in the future as this evolves we may accidentally introduce a regression in our test coverage without noticing.  Unfortunately there's no great way to test for *why* schema validation failed so the test could check that it fails for the right reason, so I don't see a good way out of this at the moment.

Reply:
I guess instead of having the schema validation just check the JSON string of the found schema against the expected schema object, it could go field-by-field and check that the schema matches exactly what we expect, throwing a specific error message for each type of failure from each type of schema mismatch, then tests could check for the specific failure message it expects.

I don't really think that's worth it to do at the moment though.  It would be better to find (or build) a tool to do that type of validation in a generic and reusable way.